mixer/pkg/server/server_test.go

```diff
@@ -0,0 +1,225 @@
+	"istio.io/istio/mixer/pkg/pool"
+	mixerRuntime "istio.io/istio/mixer/pkg/runtime"
+	"istio.io/istio/mixer/pkg/template"
+)
+
+const (
+	globalCfg = `
+apiVersion: "config.istio.io/v1alpha2"
+kind: attributemanifest
+metadata:
+  name: istio-proxy
+  namespace: default
+spec:
+    attributes:
+      source.name:
+        value_type: STRING
+      target.name:
+        value_type: STRING
+      response.count:
+        value_type: INT64
+      attr.bool:
+        value_type: BOOL
+      attr.string:
+        value_type: STRING
+      attr.double:
+        value_type: DOUBLE
+      attr.int64:
+        value_type: INT64
+---
+`
+	serviceCfg = `
+apiVersion: "config.istio.io/v1alpha2"
+kind: fakeHandler
+metadata:
+  name: fakeHandlerConfig
+  namespace: istio-system
+
+---
+
+apiVersion: "config.istio.io/v1alpha2"
+kind: samplereport
+metadata:
+  name: reportInstance
+  namespace: istio-system
+spec:
+  value: "2"
+  dimensions:
+    source: source.name | "mysrc"
+    target_ip: target.name | "mytarget"
+
+---
+
+apiVersion: "config.istio.io/v1alpha2"
+kind: rule
+metadata:
+  name: rule1
+  namespace: istio-system
+spec:
+  selector: match(target.name, "*")
+  actions:
+  - handler: fakeHandlerConfig.fakeHandler
+    instances:
+    - reportInstance.samplereport
+
+---
+`
+)
+
+// createClient returns a Mixer gRPC client, useful for tests
+func createClient(addr net.Addr) (mixerpb.MixerClient, error) {
+	conn, err := grpc.Dial(addr.String(), grpc.WithInsecure())
+	if err != nil {
+		return nil, err
+	}
+
+	return mixerpb.NewMixerClient(conn), nil
+}
+
+func TestBasic(t *testing.T) {
+	a := NewArgs()
+	a.APIPort = 0
+	a.ConfigAPIPort = 0
+	a.MonitoringPort = 0
+	a.GlobalConfig = globalCfg
+	a.ServiceConfig = serviceCfg
+
+	s, err := New(a)
+	if err != nil {
+		t.Fatalf("Unable to create server: %v", err)
+	}
+
+	err = s.Close()
+	if err != nil {
+		t.Errorf("Got error during Close: %v", err)
+	}
+}
+
+func TestClient(t *testing.T) {
+	a := NewArgs()
+	a.APIPort = 0
+	a.MonitoringPort = 0
+	a.ConfigAPIPort = 0
+	a.GlobalConfig = globalCfg
+	a.ServiceConfig = serviceCfg
+
+	s, err := New(a)
+	if err != nil {
+		t.Fatalf("Unable to create server: %v", err)
+	}
+
+	go func() {
+		_ = s.Serve()
+	}()
+
+	c, err := createClient(s.Addr())
+	if err != nil {
+		t.Errorf("Creating client failed: %v", err)
+	}
+
+	req := &mixerpb.ReportRequest{}
+	_, err = c.Report(context.Background(), req)
+
+	if err != nil {
+		t.Errorf("Got error during Report: %v", err)
+	}
+
+	err = s.Close()
+	if err != nil {
+		t.Errorf("Got error during Close: %v", err)
+	}
+}
+
+func TestErrors(t *testing.T) {
+	a := NewArgs()
+	a.APIWorkerPoolSize = -1
+	a.GlobalConfig = globalCfg
+	a.ServiceConfig = serviceCfg
+
+	s, err := New(a)
+	if s != nil || err == nil {
+		t.Errorf("Got success, expecting error")
+	}
+
+	a = NewArgs()
+	a.APIPort = 0
+	a.MonitoringPort = 0
+	a.ConfigAPIPort = 0
+	a.GlobalConfig = globalCfg
+	a.ServiceConfig = serviceCfg
+
+	for i := 0; i < 20; i++ {
```

Comment:
curious: why start 20, if we only have 7 cases ?

Reply:
Laziness. I can add cases without fussing with the loop count.